woocommerce/woocommerce
# Make `WC_Cache_Helper::geolocation_ajax_get_location_hash` case-insensitive, add tests

### Changes proposed in this Pull Request:

- #### Make `WC_Cache_Helper::geolocation_ajax_get_location_hash` case-insensitive

I was using the location hash to cache pages, and I noticed that city and state fields are case sensitive when calculating the hash. This reduces the number of cache hits as for example `London` and `london` result in different location hashes.

Side effect: This PR will invalidate any existing location hashes. I was contemplating on normalizing the hash inputs as they would most commonly be input by the user, like `ucwords( $customer->get_billing_city() )`. But since the method's intended usage is caching and the hashes may also have varied previously depending on user input, any functionality depending on this should expect the hashes to be able to vary.

- #### Add tests for `WC_Cache_Helper::geolocation_ajax_get_location_hash`



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This change is covered by a new unit test. You can run this with: 

```
pnpm --filter="@woocommerce/plugin-woocommerce" run test:php:env --filter="WC_Cache_Helper_Tests"`
```

Beyond that, you can add the following snippet to a location such as `wp-content/mu-plugins/test-pr-45329.php` (remove after testing):

```php
<?php

add_filter( 'woocommerce_customer_get_billing_city', function ( $value ) {
	if ( isset( $_GET['change-city'] ) ) {
		return esc_html( urldecode( $_GET['change-city'] ) );
	}
} );
```

With that in place:

1. Visit **WooCommerce ‣ Settings ‣ General** and set **Default customer location** to **Geolocate (with page caching support)**. Save.
2. Navigate to a frontend product page.
3. You should see something like `?v=827f0f7e4145` in the URL.
4. Overwrite this with `?change-city=copenhagen` (city name is deliberately lowercase), and hit enter (load the new URL).
5. The URL should become similar to `?change-city=copenhagen&v=1422a7be729d` (the exact hash may be different).
6. Repeat, but capitalize the city name `?change-city=Copenhagen&v=1422a7be729d` then hit enter/load the new page. This time, the hash should stay constant.
7. Repeat, but supply a completely new city name (example: `change-city=Brugges`) and hit enter. Once the new page loads, you should see we have a different hash such as `v=425f0f7e4145`.



### Changelog entry

✍🏼 *Changelog automation is being problematic here, so will update manually [-Barry]*
